CI: limit concurrency for PR builds but not for master

fingolfin opened the pull request:
The previous change which limited the number of concurrent builds
for PRs to 1 also caused builds on the master branch to become
sequential, which is not what we want. This patch fixes this.

Right now, the fix is only applied to master -- so if people work on branches directly here, then these still will be build sequentially (but only the currently running build, plus the most recent one, are being run). I think that's fine -- the one reason we want `master` to build all commits is to allow us to pinpoint which merge introduce a regression, if any.

CC @thofma @benlorenz @fieker

fingolfin committed:
CI: limit concurrency for PR builds but not for master

The previous change which limited the number of concurrent builds
for PRs to 1 also caused builds on the master branch to become
sequential, which is not what we want. This patch fixes this.

benlorenz commented:
These changes seem reasonable, I think there are very few cases of people working extensively on branches here without a corresponding PR.

But I noticed one more thing with the cancel-in-progress not behaving as I expected:
For pushes to a branch in _this_ repo (instead of a fork) where a PR already exists the current CI.yml will trigger two runs, one for the branch and one for the PR.
Due to the last line
https://github.com/oscar-system/Oscar.jl/blob/de98de6e707b0f15f2156ec587b8b41b1761a9d1/.github/workflows/CI.yml#L70
only the one for the branch really runs. But then the cancel-in-progress doesn't really cancel these running builds since these are branch-builds and not PR-builds.
(Also this will cause the CI to run directly on the branch instead of a merge commit with master)

fingolfin commented:
@benlorenz OK, that's a valid concern, but separate from what this PR does and what the auto cancelation is about, right?

We could solve this by simply asking that nobody push branches to *this* repository; just push them to your own fork (that's what I always do). Or else we could change the CI to stop building branches here other than `master` -- but I am not sure if some people rely on pushing branches here, and waiting for their CI results, *before* opening a PR... but for that I am tempted to just say "if you need to do that, push the branch to your personal fork"

fingolfin commented:
Oh and of course this is a serious concern and explains some "mysterious" problems we saw in the past:
> (Also this will cause the CI to run directly on the branch instead of a merge commit with master)

CC @thofma

fingolfin commented:
Thinking some more about it: this PR here only brings back concurrency for the *master* branch. So, what we could do is this: we could remove the line
```
 if: ${{ !(github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository) }} 
```
and simply let both the PR and the branch CI builds continue. Those for the PR will be limited to one at a time; those for the branch will also be limited to one at a time (because my hack introduced in this PR is restricted to `master`), albeit working slightly differently than for PRs: for PRs, an update of the PR immediately cancels any previously running job, and starts a new one. For branches, the currently running job is allowed to finish; but any queued updated builds for the branch are canceled. Meaning once the current job finishes, then only the one for the most recent version of the branch runs. Which I think is just fine. So we still get two instead of one build, but not more. The only *good* way to avoid this that I know is to disallow builds for non-master branches entirely. But then we'd have to code it carefully so that it only applies to *this* repository, not personal forks, else those will also be blocked from testing alternate branches... *sigh*

fingolfin committed:
CI: don't kill PR CI runs for our own branches anymore

Now that we have the concurrency feature, these additional builds are not as
problematic as they used to be anymore: there can be at most one build per PR
and one per branch (except for the master branch). So there will be at most
two parallel builds for each such branch with an active PR.

On the other hand, disabling the PR CI builds was problematic, because they
actually differ from the branch CI builds: the former are performed with a
temporary merge commit of the branch into master; the latter with just the tip
of the branch. But we really want and need the former behavior to be sure
a merge won't break master.

fingolfin committed:
CI: rename the docu job

The short name is referenced in the branch protection rules, and just 'build'
is not a very clear name for a job there.

fingolfin commented:
I pushed a corresponding commit (and another to rename the docu CI job for easier referencing in the branch protection settings).

If you think this is fine, please approve it so it can auto-merge.

benlorenz reviewed (approved):
> @benlorenz OK, that's a valid concern, but separate from what this PR does and what the auto cancelation is about, right?

Yes, that was not really related to this PR, I was just hoping you had a suggestion how to deal with that as well.

> We could solve this by simply asking that nobody push branches to _this_ repository; just push them to your own fork (that's what I always do).

Documentation previews don't work for PRs from forks, so this would also be problematic.


The new rules look good to me, thanks! We can let this run for a while and if there are long delays we can have another look later.
